_posts/2019-08-17-intro-to-reports.adoc

```diff
@@ -0,0 +1,418 @@
+
+image::intro-to-reports/visit-by-period-date-range-creteria.png[align="center", link="images/intro-to-reports/visit-by-period-date-range-creteria.png"]
+
+For each condition an external parameter will be created so that the concrete values can be inserted programmatically or alternatively by the user through a UI.
+
+=== Selecting Entity Attributes of the Visit Entity
+
+The next step in the wizard is to define the attributes of the base entity (Visit) and its associated entities that should be part of the report. The selection will influence on the one hand the `SELECT` part of the JPQL statement that is used to fetch the data. On the other hand it influences the initial version of the columns within the template, that is generated for this report when finishing the report wizard.
+
+For the `recent-visits` Report, the following attribute should be selected:
+
+* Visit Date: `visitDate`
+* Visit Description: `description`
+* Information about the associated Pet
+** the Pets' name: `pet.name`
+** the Pets' Identification Number: `pet.identificationNumber`
+** Type of the Pet: `pet.type.name`
+** The city where the Pet lives (expressed through the address of its owner): `pet.owner.city`
+
+After selecting the attributes through the UI wizard, the output template file type can be configured. In this example, the default settings are applicable.
+
+With that the report wizard is finished and the report configuration is shown for further adjustments.
+
+=== Providing External Parameters Programmatically
+
+For the use case of the `recent-visits` report the generic UI is not exposed to the user. Instead the parameters are passed in by the API call to run the report.
+
+
+In the `patient-record` report, the report screen connection was configured in the report configuration.
+
+There is an alternative approach on how to trigger report execution, which will be leveraged in this example.
+
+In this case the report instance is selected programmatically (through the `system code` attribute) and directly passed into the reports API in order to execute the report and download the resulting document.
+
+The complete listing for the triggering of the `recent-visits` report shows the main steps of:
+
+1. identifying the last month (externalized logic of relative date)
+2. loading the report instance to execute
+3. triggering the report execution and passing in the last month as a parameter
+
+
+.VisitBrowse.java
+[source,java]
+----
+
+public class VisitBrowse extends StandardLookup<Visit> {
+
+    @Inject
+    protected ReportGuiManager reportGuiManager;
+
+    @Inject
+    protected DataManager dataManager;
+
+    @Inject
+    protected TimeSource timeSource;
+
+    @Subscribe("visitsTable.lastMonthReport")
+    protected void onVisitsTableLastMonthReport(Action.ActionPerformedEvent event) {
```

Comment:
Please split lines manually in this code block, as some lines do not fit the resulting width and are wrapped automatically, which looks ugly.

Reply:
👍